common/rpc/factory.go

```diff
@@ -34,19 +36,32 @@ import (
 
 	"github.com/uber/cadence/common/log"
 	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/membership"
 )
 
-const defaultGRPCSizeLimit = 4 * 1024 * 1024
+const (
+	defaultGRPCSizeLimit = 4 * 1024 * 1024
+	factoryComponentName = "rpc-factory"
+)
 
-// Factory is an implementation of common.RPCFactory interface
-type Factory struct {
+// Factory is an implementation of rpc.Factory interface
+type FactoryImpl struct {
 	maxMessageSize int
 	channel        tchannel.Channel
 	dispatcher     *yarpc.Dispatcher
+	outbounds      *Outbounds
+	logger         log.Logger
+	serviceName    string
+	wg             sync.WaitGroup
+	ctx            context.Context
+	cancelFn       context.CancelFunc
```

Comment:
tbh I'm never sure how I feel about "context only used as a stop signal", particularly when it's 100% internal.

1: it's kinda a weird API for that.  though it is simple so I do get it.

and 2: when it's internal-only like this, the lifecycle generally does _not_ support passing in a context to the constructor, because it doesn't expect to be canceled externally.  which means we have code that _looks like_ it's designed for external cancels, but likely misbehaves if that's done, which is a bit of a landmine for later refactoring.

Reply:
IMO this is one of the better ways to implement a daemon. It standardizes stop signal propagation via calling `cancelFn` and listening `<-ctx.Done()` in main loop. 
Handling the context passed in the exposed methods is an orthogonal issue I think. Doesn't apply here. Let's discuss offline about lifecycle management since it is not realistic to figure out the "way to go" in this PR's comments

Reply:
weakly, fwiw, I'd +1 to Steven's comment, the use of context here confuses me a bit